gradio/chat_interface.py

```diff
@@ -397,68 +403,100 @@ def _display_input(
         history.append([message, None])
         return history, history
 
-    def _submit_fn(
+    async def _submit_fn(
         self,
         message: str,
         history_with_input: list[list[str | None]],
         *args,
         **kwargs,
     ) -> tuple[list[list[str | None]], list[list[str | None]]]:
         history = history_with_input[:-1]
-        response = self.fn(message, history, *args, **kwargs)
+        if self.is_async:
+            response = await self.fn(message, history, *args, **kwargs)
+        else:
+            response = self.fn(message, history, *args, **kwargs)
         history.append([message, response])
         return history, history
 
-    def _stream_fn(
+    async def _stream_fn(
         self,
         message: str,
         history_with_input: list[list[str | None]],
         *args,
         **kwargs,
-    ) -> Generator[tuple[list[list[str | None]], list[list[str | None]]], None, None]:
+    ) -> AsyncGenerator:
         history = history_with_input[:-1]
         generator = self.fn(message, history, *args, **kwargs)
         try:
-            first_response = next(generator)
+            if self.is_async:
+                first_response = await async_iteration(generator)
+            else:
+                first_response = next(generator)
```

Comment:
Since we're switching everything to coroutines, any possible blocking code that's not async needs to be run in a separate thread otherwise we'll block the event loop. So for example use `anyio.to_thread.run_sync` to run the non-async code. 

Might be better to `utils.SyncToAsyncIterator` for the generators to match what we do in `blocks`. 

Otherwise this looks good though!

Reply:
Ah nice I was going to ask why we do `utils.SyncToAsyncIterator` in `blocks.py`